src/APIClient/index.tsx

```diff
@@ -77,8 +82,42 @@ class APIClient implements APIClientInterface {
         endpoint: string,
         fileUrl: string,
         options?: UploadRequestOptions
-    ): Promise<ClientResponse> =>
-        NativeAPIClient.upload(this.baseUrl, endpoint, fileUrl, options);
+    ): ProgressPromise<ClientResponse> => {
+        const taskId = generateUploadTaskId();
+        const promise: ProgressPromise<ClientResponse> = new Promise(
+            (resolve, reject) => {
+                const uploadSubscription = Emitter.addListener(
+                    "NativeClient-UploadProgress",
+                    (e: UploadProgressEvent) => {
+                        if (e.taskId === taskId && promise.onProgress) {
+                            promise.onProgress(e.fractionCompleted);
+                        }
+                    }
+                );
+
+                NativeAPIClient.upload(
+                    this.baseUrl,
+                    endpoint,
+                    fileUrl,
+                    taskId,
+                    options
+                )
+                    .then((response) => resolve(response))
+                    .catch((error) => reject(error))
+                    .finally(() => {
+                        uploadSubscription.remove();
+                        delete promise.progress;
+                    });
+            }
+        );
+
+        promise.progress = (fn) => {
+            promise.onProgress = fn;
+            return promise;
+        };
+
+        return promise;
+    };
```

Comment:
Just following on from the top comment; I think it would simplify this method if we just handled it via a callback? The other option is also to emit the event, but offer a helper that can listen in? Or both?